src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,90 +541,431 @@ internal override void Complete()
+                string message = StringUtil.Format(ParserStrings.ProgramFailedToExecute,
+                    this.NativeCommandName, exceptionToRethrow.Message,
+                    this.Command.MyInvocation.PositionMessage);
+                ApplicationFailedException appFailedException = new ApplicationFailedException(message, exceptionToRethrow);
+
+                // There is no need to set this exception here since this exception will eventually be caught by pipeline processor.
+                // this.commandRuntime.PipelineProcessor.ExecutionFailed = true;
+
+                throw appFailedException;
+            }
+        }
+
+        private class ProcessOutputHandler
+        {
+            private int _refCount;
+            private BlockingCollection<ProcessOutputObject> _queue;
+            private bool _isFirstOutput;
+            private bool _isFirstError;
+            private bool _isXmlCliOutput;
+            private bool _isXmlCliError;
+            private string _path;
+
+            public ProcessOutputHandler(Process process, BlockingCollection<ProcessOutputObject> queue)
+            {
+                Debug.Assert(process.StartInfo.RedirectStandardOutput || process.StartInfo.RedirectStandardError, "Caller should redirect at least one stream");
+                _refCount = 0;
+                _path = process.StartInfo.FileName;
+                _queue = queue;
+
+                // we incrementing refCount on the same thread and before running any processing
+                // so it's safe to do it without Interlocked.
+                if (process.StartInfo.RedirectStandardOutput) { _refCount++; }
+                if (process.StartInfo.RedirectStandardError) { _refCount++; }
+
+                // once we have _refCount, we can start processing
+                if (process.StartInfo.RedirectStandardOutput)
+                {
+                    _isFirstOutput = true;
+                    _isXmlCliOutput = false;
+                    process.OutputDataReceived += new DataReceivedEventHandler(OutputHandler);
+                    process.BeginOutputReadLine();
+                }
+
+                if (process.StartInfo.RedirectStandardError)
+                {
+                    _isFirstError = true;
+                    _isXmlCliError = false;
+                    process.ErrorDataReceived += new DataReceivedEventHandler(ErrorHandler);
+                    process.BeginErrorReadLine();
+                }
+            }
+
+            private void decrementRefCount()
+            {
+                Debug.Assert(_refCount > 0, "RefCount should always be positive, when we are trying to decrement it");
+                if (Interlocked.Decrement(ref _refCount) == 0)
+                {
+                    _queue.CompleteAdding();
+                }
+            }
+
+            private void OutputHandler(object sender, DataReceivedEventArgs e)
```

Comment:
Minor comment: I always encourage people to use descriptive/meaningful variable names. For example, the parameter `e` here and in `ErrorHandler` below could be a bit more descriptive, like `outputReceived` and `errorReceived`.